### ISSUE TITLE
Install fails when a submodule uses commitplease

### ISSUE THREAD

**Issue by Jmclerck** [jzaefferer/commitplease#79]
As the title states I am using a project as a submodule which uses commitplease. 

The basic structure is that I have a master project which has several Docker containers. The master project pulls several submodules including one which has commitplease as a dev dependency. When the container which has commitplease in tries to do an `npm install` as build of the build container phase, the installation fails, I believe because submodules do not have a local .git folder, it is actually a reference to the submodule .git folder in the parent project.

The only workaround I have so far found is to remove commitplease as a dependency after updating the submodules and before building but this isn't an ideal situation.

Any help or guidance you could provide would be much appreciated, I have posted the nam error below for reference.

Thanks!

```
fatal: Not a git repository: ../.git/modules/customer_management_system
/root/cms/node_modules/commitplease/lib/sanitize.js:13
    throw error
    ^

Error: Command failed: git config --get core.commentchar
fatal: Not a git repository: ../.git/modules/customer_management_system

    at checkExecSyncError (child_process.js:483:13)
    at Object.execSync (child_process.js:523:13)
    at Object.<anonymous> (/root/cms/node_modules/commitplease/lib/sanitize.js:6:21)
    at Module._compile (module.js:556:32)
    at Object.Module._extensions..js (module.js:565:10)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
    at Module.require (module.js:483:17)
    at require (internal/module.js:20:19)
npm info lifecycle commitplease@2.7.2~install: Failed to exec install script
npm WARN cms@0.1.0 No repository field.
npm ERR! Linux 4.4.15-moby
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v6.4.0
npm ERR! nam  v3.10.3
npm ERR! code ELIFECYCLE

npm ERR! commitplease@2.7.2 install: `node install`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the commitplease@2.7.2 install script 'node install'.
npm ERR! Make sure you have the latest version of node.js and nam installed.
npm ERR! If you do, this is most likely a problem with the commit please package
npm ERR! not with nam itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node install
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs commit please
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls commit please
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /root/cms/npm-debug.log
```

**Comment by all3fox** [comment]
@Jmclerck thank you for the issue! Here is a reduced reproduction case for it:

```
mkdir project && cd project
git init && npm init --yes
git submodule add https://github.com/jquery/jquery
```
Now do three things to `jquery/package.json`:
 1. remove the `husky` devDependency
 1. bump to `"commitplease": "2.7.2"`
 1. change the line `"nohook": true` to `"nohook": false`

and continue:

```
cd jquery
npm install 2> errlog
```
Here are the [contents of errlog](https://gist.github.com/all3fox/9de0f694aa1b2d4bc24a97d2cb4eb0d6), it fails at `install.js` and not at `validate.js` as suggested by @Jmclerck. However, I believe the origin of the problem to be the same.

**Comment by all3fox** [comment]
@jzaefferer I have not had much experience with `git submodules`. So far, [this doc](https://git-scm.com/docs/git-submodule) gives me the impression that a programmer is not really supposed/expected to change the submodule content  (text-search for "cannot modify") and the whole point is that the history of supermodule and submodule are separate.

Yet another doc goes to great lengths about how to publish local changes to git submodule ([roughly here](https://git-scm.com/book/en/v2/Git-Tools-Submodules#Working-on-a-Project-with-Submodules)).

My opinion is that we should not invade the supermodule if we are a devDependency of a submodule. But maybe that is just my laziness talking :)

When `commitplease` is a devDependency of a submodule, the `.git` of the submodue is an ordinary file that contains a relative path to the supermodule. I have not found a documented guarantee for that yet, it is just what I see in console. So I am thinking of the following workaround: if we see such a file, do not install. What do you think?

**Comment by jzaefferer** [comment]
No installing the commit hook when being installed in a submodule sounds good to me. That particular check seems fine, too.

**Comment by Jmclerck** [comment]
Hi @jzaefferer + @all3fox thanks for looking into this so quickly. As a user I would be happy with the change of commit hooks not being installed if the project is a submodule. If that is your decision I look forward to those changes being made and if there is anything I can do to assist you please let me know. Thanks again!

**Comment by all3fox** [comment]
@Jmclerck since you are using the latest `2.7.2`, I assume you actually have access to the submodule that fails, right? If so, could you please hard-code `"commitplease": 2.7.3-0` in the `package.json` of that submodule and tell us if the problem is fixed with that beta release?

I have tested it locally, works for me.

**Comment by Jmclerck** [comment]
Hi @all3fox, I have tried the `beta` version as requested, everything appears to work as expected when installing on a local machine, however the issue still exists when my Docker container tries to build and executes `npm install`. The error is still the same as my original post but for reference:
```
> commitplease@2.7.3-0 install /root/cms/node_modules/commitplease
> node install

fatal: Not a git repository: ../.git/modules/customer_management_system
/root/cms/node_modules/commitplease/lib/sanitize.js:13
    throw error
    ^

Error: Command failed: git config --get core.commentchar
fatal: Not a git repository: ../.git/modules/customer_management_system

    at checkExecSyncError (child_process.js:483:13)
    at Object.execSync (child_process.js:523:13)
    at Object.<anonymous> (/root/cms/node_modules/commitplease/lib/sanitize.js:6:21)
    at Module._compile (module.js:556:32)
    at Object.Module._extensions..js (module.js:565:10)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
    at Module.require (module.js:483:17)
    at require (internal/module.js:20:19)
npm info lifecycle commitplease@2.7.3-0~install: Failed to exec install script
```

I noticed my local and Docker versions were out of sync so I upgraded both to the latest node `v6.5.0` and npm `v3.10.3`, but still no luck.

**Comment by all3fox** [comment]
The idea behind this patch is that the require chain `install.js` -> `index.js` -> `sanitize.js` triggers the code inside `sanitize.js` which fails on line 13. Moving that code inside the `sanitize` function means that now it will not be run on require.

@Jmclerck could you try your Docker routine again, using `beta` as the version for commitplease? Also, if that is not too much trouble, maybe you could tell me how to set up a Docker environment similar to yours?

**Comment by Jmclerck** [comment]
Hi @all3fox I just tried the `beta` version again inside my Docker container and everything worked perfectly. Console output from `npm install` contained this message which I believe is the desired effect:

```
> commitplease@2.7.3-1 install /root/cms/node_modules/commitplease
> node install

Will not install in a git submodule
```

Thanks again 👍

**Comment by all3fox** [comment]
Hi @Jmclerck, I am glad to hear that `commitplease@beta` is working for you. Could you try `commitplease@next` though? Just change your package.json to `"commitplease": "next"` and it should install. This `next` is the version likely to become `commitplease@2.7.3`

**Comment by Jmclerck** [comment]
Hi @all3fox, I'm on holiday this week so I won't be able to test this until next Monday, I hope that isn't a problem. I'll get back to you as soon as I have tested `"commitplease": "next"`, thanks

**Comment by Jmclerck** [comment]
Hi @all3fox, I am back at work today after my holiday (boo) but I have tried the `next` version of `commitplease` and again the install completed successfully, this is using `node@v4.5.0` and `npm@2.15.9`, I have been unable to test with `node@6.x` and `npm@3.x` because of an unrelated Docker issue, but I think I will need to raise another issue for that. As for submodules everything seems to work correctly! 👍

**Comment by all3fox** [comment]
thanks @Jmclerck, it is now published as `2.7.3`, make sure to eventually stop using `beta` and `next` tags as they will be pointing at unstable versions, also reopen if necessary :)
